File: dsa-learning/dsa-python-code/com/dsa/leetcode/labuladong/binary_tree/P104_MaximunDepthOfBinaryTree.py
# author: tobebetter9527
# since: 2023/11/6 10:07
from typing import Optional

from com.dsa.leetcode.labuladong.binary_tree import TreeNode


class Solution:

    def maxDepth(self, root: Optional[TreeNode]) -> int:
        if not root:
            return 0
        len1 = self.maxDepth(root.left)
        len2 = self.maxDepth(root.right)
        return max(len2, len1) + 1

File: dsa-learning/dsa-java-code/src/main/java/com/dsa/leetcode/labuladong/binary_tree/P104_MaximunDepthOfBinaryTree.java
package com.dsa.leetcode.labuladong.binary_tree;

public class P104_MaximunDepthOfBinaryTree {
    int maxDepth = 0;
    int depth = 0;

    /**
     * 遍历，回溯思想
     * @param root
     * @return
     */
    public int maxDepth(TreeNode root) {
        traverse(root);
        return maxDepth;
    }

    private void traverse(TreeNode root) {
        if (root == null) {
            return;
        }
        depth++;
        traverse(root.left);
        traverse(root.right);
        maxDepth = Math.max(maxDepth, depth);
        depth--;
    }

    /**
     * 分解问题思想
     * @param root
     * @return
     */
    public int maxDepth2(TreeNode root) {
        if (root == null) {
            return 0;
        }
        int depth1 = maxDepth2(root.left);
        int depth2 = maxDepth2(root.right);
        return Math.max(depth2,depth1) + 1;
    }

}
